contracts/specs/NFTSwapper/NFTSwapperSpec.md

```diff
@@ -0,0 +1,163 @@
+
+This contract is immutable with no owner. No NFTs or ETH are ever held by this contract, instead it just facilitates the permissionless swap process. 
+
+---
+
+- [NFTSwapper Spec](#nftswapper-spec)
+  - [Function Interface](#function-interface)
+    - [`constructor`](#constructor)
+    - [`swapPrice`](#swapprice)
+    - [`swap`](#swap)
+  - [Swap Price Curve](#swap-price-curve)
+    - [Constant Price Curve](#constant-price-curve)
+    - [Output NFT age based curve](#output-nft-age-based-curve)
+    - [Input and Output NFT age difference based curve](#input-and-output-nft-age-difference-based-curve)
+  - [Approval Considerations](#approval-considerations)
+
+## Function Interface
+
+### `constructor`
+
+```solidity
+constructor(
+    address swapPool, 
+    address nft, 
+    address feeRecipient, 
+    PriceCurveParams priceCurveParams
+)
+```
+
+* `swapPool`: the address which holds the swappable nft's, note that the `swapPool` must approve this contract for it's NFT's (or a subset of them, see [Approval Considerations](#approval-considerations)).
+* `nft`: the address of the ERC721 token that is being swapped.
+* `feeRecipient`: the address where swap fees will be sent. 
+* `priceCurveParams`: the parameters for the price curve (see [Swap Price Curve](#swap-price-curve)).
+ 
+### `swapPrice`
+
+```solidity
+function swapPrice(
+    uint256 inputTokenId, 
+    uint256 outputTokenId
+) pure returns (uint256)
+```
+
+Pure function to get the price in Wei to perform a swap from `inputTokenId` to `outputTokenId` based on [swap price curve](#swap-price-curve).
+
+### `swap`
+
+```solidity
+function swap(uint256 inputTokenId, uint256 outputTokenId) payable 
+```
+
+Swaps the `inputTokenId` from `msg.sender` with the `outputTokenId` from the `swapPool`. Sends the swap fee to the `feeRecipient`.  
+
+Requirements:
+* This contract must be approved to transfer the `nft` with `inputTokenId` on behalf of `msg.sender`.
+* This contract must be approved to transfer the `nft` with `outputTokenId` on behalf of the `swapPool`
+* Swap fee of `swapPrice(inputTokenId, outputTokenId)` Wei must be provided 
+
+
+## Swap Price Curve 
+
+The swap price is determined by the curve defined below. The `PriceCurveParams` are configured on deployment and cannot be changed.
+
+> Currently 3 different options are being considered, only 1 will be selected once finalized
+> - [Constant Price Curve](#constant-price-curve)
+> - [Output NFT age based curve](#output-nft-age-based-curve)
+> - [Input and Output NFT age difference based curve](#input-and-output-nft-age-difference-based-curve)
+
+### Constant Price Curve
+
+Simplest price curve where every swap is the same `floor` price.
+
+```solidity
+struct PriceCurveParams {
+    uint256 floor;
+}
+
+PriceCurveParams immutable priceCurveParams;
+
+function swapPrice() pure returns (uint256) {
+    return priceCurveParams.floor;
+}
+```
+
+![](img/constant-price-curve.png)
+
+Pros:
+* Simplicity
```

Comment:
my preference is constant because simple is good

Reply:
would we ever want this to be updatable? Probably impossible if we are making the contract ownerness

Reply:
Yea, been thinking about this. 

Question: Should the priceCurveParams be updatable by the DAO

Probably yes, original thinking was no because we can just:
1. redeploy with new param
2. revoke approval on existing deployment
3. Set approval for new deployment

But, if we don't do an approveAll this gets trickier/annoying to do. Open to either, pro of no owner is less surface area in general, its just a single external `swap` function.

Reply:
In terms of constant vs an ID based price curve. 

I agree constant is nice because simple = better.

But... based on community feedback we've received, I think tokenId difference based pricing will gain more community alignment by maintaining (at least some) non-fungibility.